src/controllers/hid/hidiothread.cpp

```diff
@@ -0,0 +1,234 @@
+    if (bytesRead == m_lastPollSize &&
+            memcmp(pCurrentBuffer, pPreviousBuffer, bytesRead) == 0) {
+        return;
+    }
+    // Cycle between buffers so the memcmp above does not require deep copying to another buffer.
+    m_pollingBufferIndex = (m_pollingBufferIndex + 1) % kNumBuffers;
+    m_lastPollSize = bytesRead;
+
+    // Convert array of bytes read in a JavaScript compatible return type, this is emitted as deep-copy, for thread safety.
+    // This eexecute callback function in JavaScript mapping and print to stdout in case of --controllerDebug
+    emit receive(QByteArray(reinterpret_cast<const char*>(pCurrentBuffer),
+                         bytesRead),
+            mixxx::Time::elapsed());
+}
+
+QByteArray HidIoThread::getInputReport(unsigned int reportID) {
+    auto startOfHidGetInputReport = mixxx::Time::elapsed();
+    auto lock = lockMutex(&m_HidDeviceMutex);
+    int bytesRead;
+
+    m_pPollData[m_pollingBufferIndex][0] = reportID;
+    // FIXME: implement upstream for hidraw backend on Linux
+    // https://github.com/libusb/hidapi/issues/259
+    bytesRead = hid_get_input_report(m_pHidDevice, m_pPollData[m_pollingBufferIndex], kBufferSize);
+    qCDebug(m_logInput) << bytesRead << "bytes received by hid_get_input_report"
+                        << m_deviceInfo.formatName() << "serial #"
+                        << m_deviceInfo.serialNumber()
+                        << "(including one byte for the report ID:"
+                        << QString::number(static_cast<quint8>(reportID), 16)
+                                   .toUpper()
+                                   .rightJustified(2, QChar('0'))
+                        << ") - Needed: "
+                        << (mixxx::Time::elapsed() - startOfHidGetInputReport)
+                                   .formatMicrosWithUnit();
+
+    if (bytesRead <= kReportIdSize) {
+        // -1 is the only error value according to hidapi documentation.
+        // Otherwise minimum possible value is 1, because 1 byte is for the reportID,
+        // the smallest report with data is therefore 2 bytes.
+        DEBUG_ASSERT(bytesRead <= kReportIdSize);
+        return QByteArray();
+    }
+
+    // Convert array of bytes read in a JavaScript compatible return type, this is returned as deep-copy, for thread safety.
+    // For compatibility with HidController::processInputReport, the reportID prefix is included added here
+    return QByteArray(reinterpret_cast<char*>(m_pPollData[m_pollingBufferIndex]), bytesRead);
+}
+
+void HidIoThread::sendOutputReport(const QByteArray& data, unsigned int reportID) {
+    {
+        auto lock = lockMutex(&m_HidDeviceMutex);
+        if (m_outputReports.find(reportID) == m_outputReports.end()) {
+            std::unique_ptr<HidIoReport> pNewOutputReport;
+            m_outputReports[reportID] = std::make_unique<HidIoReport>(
+                    reportID, m_pHidDevice, std::move(m_deviceInfo));
```

Comment:
ah ok so you need the device info with every report? again this std::moving is not going to do what you want and will always be making copies anyway.  How about pass it as a shared pointer?  I assume the lifetime of the report is guaranteed to be shorter than the lifetime of the deviceinfo.

Reply:
I changed this to a shared_ptr !